arch/arm/core/aarch64/arm_mmu.c

```diff
@@ -0,0 +1,464 @@
+/*
+ * Copyright 2019 Broadcom
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <init.h>
+#include <kernel.h>
+#include <arch/arm/aarch64/cpu.h>
+#include <arch/arm/aarch64/arm_mmu.h>
+#include <linker/linker-defs.h>
+#include <sys/util.h>
+
+/* Set below flag to get debug prints */
+#define MMU_DEBUG_PRINTS	0
+/* To get prints from MMU driver, it has to initialized after console driver */
+#define MMU_DEBUG_PRIORITY	70
+
+#if MMU_DEBUG_PRINTS
+/* To dump page table entries while filling them, set DUMP_PTE macro */
+#define DUMP_PTE		0
+#define MMU_DEBUG(fmt, ...)	printk(fmt, ##__VA_ARGS__)
+#else
+#define MMU_DEBUG(...)
+#endif
+
+/* We support only 4kB translation granule */
+#define PAGE_SIZE_SHIFT		12U
+#define PAGE_SIZE		(1U << PAGE_SIZE_SHIFT)
+#define XLAT_TABLE_SIZE_SHIFT   PAGE_SIZE_SHIFT /* Size of one complete table */
+#define XLAT_TABLE_SIZE		(1U << XLAT_TABLE_SIZE_SHIFT)
+
+#define XLAT_TABLE_ENTRY_SIZE_SHIFT	3U /* Each table entry is 8 bytes */
+#define XLAT_TABLE_LEVEL_MAX	3U
+
+#define XLAT_TABLE_ENTRIES_SHIFT \
+			(XLAT_TABLE_SIZE_SHIFT - XLAT_TABLE_ENTRY_SIZE_SHIFT)
+#define XLAT_TABLE_ENTRIES	(1U << XLAT_TABLE_ENTRIES_SHIFT)
+
+/* Address size covered by each entry at given translation table level */
+#define L3_XLAT_VA_SIZE_SHIFT	PAGE_SIZE_SHIFT
+#define L2_XLAT_VA_SIZE_SHIFT	\
+			(L3_XLAT_VA_SIZE_SHIFT + XLAT_TABLE_ENTRIES_SHIFT)
+#define L1_XLAT_VA_SIZE_SHIFT	\
+			(L2_XLAT_VA_SIZE_SHIFT + XLAT_TABLE_ENTRIES_SHIFT)
+#define L0_XLAT_VA_SIZE_SHIFT	\
+			(L1_XLAT_VA_SIZE_SHIFT + XLAT_TABLE_ENTRIES_SHIFT)
+
+#define LEVEL_TO_VA_SIZE_SHIFT(level) \
+				(PAGE_SIZE_SHIFT + (XLAT_TABLE_ENTRIES_SHIFT * \
+				(XLAT_TABLE_LEVEL_MAX - (level))))
+
+/* Virtual Address Index within given translation table level */
+#define XLAT_TABLE_VA_IDX(va_addr, level) \
+	((va_addr >> LEVEL_TO_VA_SIZE_SHIFT(level)) & 0x1ffULL)
```

Comment:
Instead of hardcoding `0x1ffULL` what about `XLAT_TABLE_ENTRIES - 1`? If you are going to change the size of the translation table it's better to just change one parameter.

Reply:
Will use `XLAT_TABLE_ENTRIES - 1`.